flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -72,123 +202,217 @@ int32_t vtol_follower_control_path(const float dT, const PathDesiredData *pathDe
-	float altitudeSetpoint = pathDesired->Start[2] + (pathDesired->End[2] - pathDesired->Start[2]) *
-	    bound_min_max(progress->fractional_progress, 0, 1);
+	commands_ned[1] = progress->path_direction[1] * groundspeed +
+	    progress->correction_direction[1] * error_speed;
 
-	float downError = altitudeSetpoint - positionActual.Down;
-	velocityDesired.Down = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
+	/* Limit the total velocity based on the configured value. */
+	vtol_limit_velocity(commands_ned, guidanceSettings.HorizontalVelMax);
+
+	commands_ned[2] = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
 		-guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
 
+	VelocityDesiredData velocityDesired;
+	VelocityDesiredGet(&velocityDesired);
+	velocityDesired.North = commands_ned[0];
+	velocityDesired.East = commands_ned[1];
+	velocityDesired.Down = commands_ned[2];
 	VelocityDesiredSet(&velocityDesired);
 
+	pathStatus.Status = PATHSTATUS_STATUS_INPROGRESS;
+	PathStatusSet(&pathStatus);
+
 	return 0;
 }
 
 /**
- * Control algorithm to stay or approach at a fixed location.
- * @param[in] dT the time since last evaluation
- * @param[in] ned The position to attempt to hold
- * This method does not attempt any particular path, simply a straight line
- * approach. The calculated velocity to attempt is stored in @ref 
- * VelocityDesired. 
- *
- * Takes in @ref PositionActual and compares it to @ref PositionDesired 
- * and computes @ref VelocityDesired
+ * Controller to maintain/seek a position and optionally descend.
+ * @param[in] dT time since last eval
+ * @param[in] hold_pos_ned a position to hold
+ * @param[in] landing whether to descend
+ * @param[in] update_status does this update path_status, or does somoene else?
  */
-int32_t vtol_follower_control_endpoint(const float dT, const float *hold_pos_ned)
-{
+static int32_t vtol_follower_control_simple(const float dT,
+	const float *hold_pos_ned, bool landing, bool update_status) {
 	PositionActualData positionActual;
 	VelocityDesiredData velocityDesired;
 	
 	PositionActualGet(&positionActual);
-	VelocityDesiredGet(&velocityDesired);
-	
-	float northError;
-	float eastError;
-	float downError;
-	float northCommand;
-	float eastCommand;
 	
-	const float cur_pos_ned[3] = { positionActual.North, positionActual.East, positionActual.Down };
+	VelocityActualData velocityActual;
+
+	VelocityActualGet(&velocityActual);
+
+	/* Where would we be in ___ second at current rates? */	
+	const float cur_pos_ned[3] = {
+		positionActual.North +
+		    velocityActual.North * guidanceSettings.PositionFeedforward,
+		positionActual.East +
+		    velocityActual.East * guidanceSettings.PositionFeedforward,
+		positionActual.Down };
+
+	float errors_ned[3];
+
+	/* Calculate the difference between where we want to be and the
+	 * above position */
+	vtol_calculate_distances(cur_pos_ned, hold_pos_ned, errors_ned, 0);
+
+	float horiz_error_mag = vtol_magnitude(errors_ned, 2);
```

Comment:
You didn't like my nicer elegant way of writing this math I put in the previous comment?
https://github.com/TauLabs/TauLabs/pull/1543#discussion_r30106680